### PR TITLE
get list of gem files from file system and .gitignore

### PR DESCRIPTION
**Pull request by jmechnich**
Currently, it is not possible to build a gem from the GitHub archives for each tag. This PR should fix that.

### CODE REVIEW

**Comment by joeyates** [comment]
Hi @jmechnich thanks for the PR.

I've commented on the Homebrew PR (Homebrew/homebrew-core#88089).

I'm not actually clear why this PR is necessary build the gem locally from the GitHub archive.

As an experiment, I downloaded the [latest tag](https://github.com/joeyates/imap-backup/releases/tag/v3.4.0) as a zip, and ran 

```shell
$ gem build imap-backup.gemspec
$ gem install imap-backup-3.4.0.gem
```

and this installed the gem.

Could you explain why these changes are necessary?

Thanks!

**Comment by jmechnich** [comment]
You will end up with an empty gem:
```
mechnich@darkstar:~/Downloads/imap-backup-3.4.0$ gem build imap-backup.gemspec 
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
WARNING:  licenses is empty, but is recommended.  Use a license identifier from
http://spdx.org/licenses or 'Nonstandard' for a nonstandard license.
WARNING:  no files specified
```
The tag archive files don't contain the `.git` directory, so `git ls-files` cannot work.

**Comment by joeyates** [comment]
Hi @jmechnich 

Thanks for your clarification.

As it is not normal practice for gems to be buildable in the way that this PR proposes, I would like to understand why this is a special case.

As I mentioned in Homebrew/homebrew-core#88089, I'm not clear why the formula doesn't simply install the gem  directly, and has to build it locally.
It looks like that is the normal practice in Homebrew - for example [here](https://github.com/Homebrew/homebrew-core/blob/cbda64c699e029697f8c0ee89052ffcfcdf589e6/Formula/mikutter.rb)

**Comment by jmechnich** [comment]
I am not an experienced Ruby coder, so I can't comment on what is supposed to be "normal".
However, I expect to be able to build a software successfully from a source archive, which failed.
I can understand though that the previous version might be more convenient when building from a cloned repo.

As for the homebrew formula, maybe installing the gem directly is an alternative way for implementing it.
Basically, I just started out with the template for a ruby build (`brew create --ruby ...`), and there are other formulas that work the same way, for example [travis.rb](https://github.com/Homebrew/homebrew-core/blob/0e4faa62f0043f5eecef1c1076b779b10ed549d0/Formula/travis.rb).
It was not a suggested change from the reviewers though, so, again, I can't really comment on what is "normal".

From my naive point of view, it seems like a good idea to get rid of the `git ls-files` in order to make the source archives sort of self-consistent.

**Comment by joeyates** [comment]
Thanks for the additional info.

I checked [the most popular gems](https://rubygems.org/stats) (+Rails), and they use 4 different ways of constructing the `files` list:

# Dir.glob + Rejection of directories + Explicit List

* [Bundler (1st)](https://github.com/rubygems/rubygems/blob/master/bundler/bundler.gemspec#L37)

# Dir() Mixing Globs and Explicit List

Example:

```ruby
Dir["CHANGELOG.md", "MIT-LICENSE", "README.rdoc", "lib/**/*"]
```

* [activesupport](https://github.com/rails/rails/blob/v6.1.4.1/activesupport/activesupport.gemspec#L20)
* [aws-sdk-core (8th)](https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-core/aws-sdk-core.gemspec#L13)
* [multi_json (9th)](https://github.com/intridea/multi_json/blob/v1.15.0/multi_json.gemspec#L9)

# Explicit list

[diff-lcs (4th)](https://github.com/halostatue/diff-lcs/blob/master/diff-lcs.gemspec#L17)
[rails, which has only one file](https://github.com/rails/rails/blob/v6.1.4.1/rails.gemspec#L21)

# `git ls` + Explicit List

[rspec-expectations (2nd)](https://github.com/rspec/rspec-expectations/blob/main/rspec-expectations.gemspec#L23)
[rspec-core (3rd)](https://github.com/rspec/rspec-core/blob/main/rspec-core.gemspec#L24)
[rspec-mocks (5th)](https://github.com/rspec/rspec-mocks/blob/main/rspec-mocks.gemspec#L24)
[rspec-support (6th)](https://github.com/rspec/rspec-support/blob/main/rspec-support.gemspec#L24)
[rspec (7th)](https://github.com/rspec/rspec-metagem/blob/main/rspec.gemspec#L24)

# Conclusion

So, of the most popular gems, only the RSpec families does things like imap-backup.

The best practice seems to be to use `Dir.glob` (or `Dir`, which is equivalent), or an explicit list, which I think adds maintenance load.

File: imap-backup.gemspec [hunk 1]
```diff
@@ -8,8 +9,9 @@ Gem::Specification.new do |gem|
   gem.authors       = ["Joe Yates"]
   gem.email         = ["joe.g.yates@gmail.com"]
   gem.homepage      = "https://github.com/joeyates/imap-backup"
-
-  gem.files         = `git ls-files`.split($OUTPUT_RECORD_SEPARATOR)
+  gem.licenses      = ["MIT"]
+  
+  gem.files         = Rake::FileList["**/*"].exclude(*File.read(".gitignore").split)
```
Comment (joeyates) [inline imap-backup.gemspec:14]: Could this use `Dir.glob`?

File: imap-backup.gemspec [hunk 2]
```diff
@@ -8,8 +9,9 @@ Gem::Specification.new do |gem|
   gem.authors       = ["Joe Yates"]
   gem.email         = ["joe.g.yates@gmail.com"]
   gem.homepage      = "https://github.com/joeyates/imap-backup"
-
-  gem.files         = `git ls-files`.split($OUTPUT_RECORD_SEPARATOR)
+  gem.licenses      = ["MIT"]
```
Comment (joeyates) [inline imap-backup.gemspec:12]: As this is a separate matter from the listing functionality, could it go in a separate commit?

**Review by joeyates — CHANGES_REQUESTED** [review]
The commit message will need to be amended to reflect the fact that .gitignore is not consulted.

The first letter [should be capitalized](https://www.git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines).

File: imap-backup.gemspec (hunk 1 above)
Comment (jmechnich) [reply imap-backup.gemspec:14]: The accepted answer to [this](https://stackoverflow.com/questions/11873294/determining-the-gems-list-of-files-for-the-specification) stackoverflow question seems to list a few sensible options.

No idea if the dependency to `Rake::FileList` is something to be avoided, but I thought it kind of elegant to re-use `.gitignore` as a generic file rejection list. If you prefer `Dir.glob` I am happy to amend the commit accordingly.

**Comment by jmechnich** [comment]
@joeyates I am not sure if this is the solution that you will be happy with but here you go. Please let me know if any important files are missing.

**Review by joeyates — APPROVED** [review]

**Comment by joeyates** [comment]
Hi @jmechnich 

I did a manual rebase and merge via the command line.

Thanks again for the PR!

**Comment by jmechnich** [comment]
@joeyates I have just noticed that `imap-backup.gemspec` still contained the now superfluous `require "rake/file_list"` which leads to your CI failing. I have amended my branch accordingly but maybe you just want to remove it. Sorry for my negligence. :(
